File: dvbgrab/backend/vlc/httpGrab.php
#!/usr/bin/php
<?php
//Grabs http stream
//Usage: httpGrab -b BEGIN_TIME -e END_TIME -i INPUT_CHANNEL -o OUTPUT_FILE
//Time format: "Y-m-d H:i:s" or any English textual datetime
//Channels: ct1, ct2, ct24, nova, prima
//Example:
// ./httpGrab.php -b '+10 seconds' -e '+100 seconds' -i nova -o test.ps

define("arg_BEGIN_TIME", "-b");
define("arg_END_TIME", "-e");
define("arg_INPUT_CHANNEL", "-i");
define("arg_OUTPUT_FILE", "-o");

define("URL_PREFIX", "http://localhost:1234/");

/**
* Parses command line options.
* @returns array of key=value pairs
*/
function parseArgs($argv) {
    $options = array();
    for ($i = 1; $i < count($argv); $i++) {
        if ($i % 2 == 0) {
            $options[$key] = $argv[$i];
        }
        else {
            $key = $argv[$i];
        }
    }
    return $options;
}

function waitForBegin($begin_time) {
    $delay = $begin_time - time();
    if ($delay > 0) {
        echo "Sleeping ${delay}s\n";
        sleep($delay);
    }
}

/**
* Dumps http stream until end_time.
* @param $input_channel channel number
* @param $output_file filename
* @param $end_time end time as number of seconds since epoch
*/
function dumpStream($input_channel, $output_file, $end_time) {
    $output = fopen($output_file, "w");
    if (!$output) {
        die("Error: cannot open output file: '$output_file'\n");
    }

    $stream_url = URL_PREFIX . $input_channel;
    $stream = fopen($stream_url, "r");
    if (!$stream) {
        die("Error: cannot open stream url: '$stream_url'\n");
    }

    $start = time();
    $length = $end_time - $start;
    $progress = 0;

    //NOTE: bigger buffer than 8192 does not take effect
    echo "Dumping $stream_url > $output_file\n";
    while (!feof($stream) && time() <= $end_time) {
        fwrite($output, fread($stream, 8192));

        $percent = 100 * (time() - $start) / $length;
        while ($percent >= $progress) {
            $progress += 1;
            echo ".";
        }
    }
    echo "\n";

    if (feof($stream)) {
        echo "Warning: end of stream\n";
    }

    fclose($output);
    fclose($stream);
}

function main($argv) {
    $options = parseArgs($argv);

    $begin_time = time();
    $end_time = time() + 24*3600;
    $input_channel = 1;
    $output_file = "/dev/null";

    if (isset($options[arg_BEGIN_TIME])) {
        $begin_time = strtotime($options[arg_BEGIN_TIME]);
    }
    if (isset($options[arg_END_TIME])) {
        $end_time = strtotime($options[arg_END_TIME]);
    }
    if (isset($options[arg_INPUT_CHANNEL])) {
        $input_channel = $options[arg_INPUT_CHANNEL];
    }
    if (isset($options[arg_OUTPUT_FILE])) {
        $output_file = $options[arg_OUTPUT_FILE];
    }

    echo "begin_time=$begin_time\n";
    echo "end_time=$begin_time\n";
    echo "input_channel='$input_channel'\n";
    echo "output_file='$output_file'\n";

    if ($begin_time >= $end_time) {
        die("Error: cannot set begin time after end time!\n");
    }

    waitForBegin($begin_time);
    dumpStream($input_channel, $output_file, $end_time);
}

//Clean start without any global variable
main($argv);

?>

File: backend/vlc/httpGrab.py
#!/usr/bin/env python
"""\
Usage: httpGrab -b BEGIN_TIME -e END_TIME -i CHANNEL_NAME -o OUTPUT_FILE
Grabs HTTP stream.
Time format: 'Y-M-D h:m:s'
Channels: nova, prima, ct1, ct2, ct24
Example:
./httpGrab -e '2005-10-26 14:40:00' -i nova -o test.mpg
"""
URL_PREFIX = "http://localhost:1234/"
PROXIES = {}
arg_BEGIN_TIME = "-b"
arg_END_TIME = "-e"
arg_CHANNEL_NAME = "-i"
arg_OUTPUT_FILE = "-o"

import sys
import time
import urllib2

def openUrl(url):
    """ Opens url or dies.
    Returns stream handler.
    """
    try:
        proxyHandler = urllib2.ProxyHandler(PROXIES)
        opener = urllib2.build_opener(proxyHandler)
        return opener.open(url)
    except:
        sys.exit("Error: cannot open url: %s, error: %s: %s"
                % (url, sys.exc_info()[0], sys.exc_info()[1]))

def dumpStream(url, outputFilename, endTime):
    stream = openUrl(url)
    output = file(outputFilename, 'w')

    start = time.time()
    length = endTime - start
    print "Dumping %s > %s for %.2fm" % (url, outputFilename, length/60.0)
    progress = 0
    bufferSize = 8*1024
    while time.time() <= endTime:
        buffer = stream.read(bufferSize)
        if len(buffer) == 0:
            print "Error: end of data, partial output: %s" % outputFilename
            break
        output.write(buffer)
        percent = 100 * (time.time() - start) / length
        while percent >= progress:
            progress += 1
            sys.stdout.write('.')
            sys.stdout.flush()
    output.close()
    stream.close()
    print "File %s is ready" % outputFilename

def waitForBegining(beginTime):
    delay = int(beginTime - time.time())
    if delay > 0:
        print "Waiting %ds" % delay
        time.sleep(delay)

def parseArgs(args):
    """ Returns 'key->value' pairs.
    """
    key = None
    options = {}
    for index, value in enumerate(args):
        if index % 2 == 0:
            options[key] = value
        else:
            key = value
    return options

def parseTime(string):
    """ Returns nuber for seconds since Epoch.
    """
    return time.mktime(time.strptime(string, '%Y-%m-%d %H:%M:%S'))

def strFromTime(seconds):
    return time.strftime('%Y-%m-%d %H:%M:%S', time.localtime(seconds))

def usage():
    print __doc__

def main():
    channelName = None
    outputFilename = "/dev/null"
    beginTime = time.time()
    endTime = beginTime + 24*3600

    options = parseArgs(sys.argv)
    if not arg_CHANNEL_NAME in options:
        usage()
        return
    else:
        channelName = options[arg_CHANNEL_NAME] 
    if arg_OUTPUT_FILE in options:
        outputFilename = options[arg_OUTPUT_FILE] 
    if arg_BEGIN_TIME in options:
        beginTime = parseTime(options[arg_BEGIN_TIME])
    if arg_END_TIME in options:
        endTime = parseTime(options[arg_END_TIME])

    beginTime = max(time.time(), beginTime)
    #print "beginTime:", strFromTime(beginTime)
    #print "endTime:", strFromTime(endTime)
    if beginTime >= endTime:
        sys.exit("Error: begin time must be before end")
    waitForBegining(beginTime)
    url = URL_PREFIX + channelName
    dumpStream(url, outputFilename, endTime)

#-----------------------------------------------------------------
main()
